File: saas/axops/src/ui/src/app/model/deployment.ts
import { Template } from './template';
import { Pod } from './pod';
import { ExternalRoute, TerminationPolicy } from '.';

export const DEPLOYMENT_STATUSES = {
    'INIT': 'Init',
    'WAITING': 'Waiting',
    'ACTIVE': 'Active',
    'ERROR': 'Error',
    'TERMINATING': 'Terminating',
    'TERMINATED': 'Terminated',
    'STOPPED': 'Stopped',
    'STOPPING': 'Stopping'
};

export const DeploymentFieldNames = {
    name: 'name',
    status: 'status',
};

export class Deployment {
    annotations: string[];
    app_generation: string;
    app_id: string;
    app_name: string;
    costid: any[];
    cpu: number;
    create_time: number;
    deployment_id: string;
    description: string;
    end_time: number;
    id: string;
    instances: Pod[] = [];
    labels: any = {};
    launch_time: number;
    mem: number;
    name: string;
    parameters: any[];
    run_time: number;
    status: string;
    status_detail: any;
    task_id: string;
    template: Template;
    template_id: string;
    user: string;
    wait_time: number;
    termination_policy: TerminationPolicy;
    jira_issues: string[];
    previous_deployment_id?: string;

    constructor(data?) {
        if (typeof data === 'object') {
            for (let key in data) {
                if (data.hasOwnProperty(key)) {
                    this[key] = data[key];
                }
            }
        }
    }

    public canStop(): boolean {
        let arr = [DEPLOYMENT_STATUSES.TERMINATED,
        DEPLOYMENT_STATUSES.TERMINATING,
        DEPLOYMENT_STATUSES.STOPPED,
        DEPLOYMENT_STATUSES.STOPPING];
        return arr.indexOf(this.status) > -1 ? false : true;
    }

    public canTerminate(): boolean {
        let arr = [DEPLOYMENT_STATUSES.TERMINATED];
        return arr.indexOf(this.status) > -1 ? false : true;
    }

    public canStart(): boolean {
        return this.status === DEPLOYMENT_STATUSES.STOPPED;
    }

    public canChangeScale(): boolean {
        return this.status !== DEPLOYMENT_STATUSES.STOPPED && this.status !== DEPLOYMENT_STATUSES.TERMINATED;
    }

    public getDeplymentLabels() {
        let l = [];
        for (let key in this.labels) {
            if (this.labels[key]) {
                l.push({ key, value: this.labels[key] });
            }
        }
        return l;
    }

    public getExternalRoutes(): ExternalRoute[] {
        let arr: ExternalRoute[] = [];
        if (this.template && this.template.external_routes && this.template.external_routes.length > 0) {
            arr = this.template.external_routes;
        }
        return arr;
    }

    public getFirstExternalRoute(): string {
        let arr: ExternalRoute[] = this.getExternalRoutes();
        return arr.length > 0 ? arr[0].dns_name : '';
    }
}

File: saas/axamm/src/applatix.io/axamm/deployment/deployment.go
package deployment

import (
	"fmt"

	"applatix.io/axdb"
	"applatix.io/axerror"
	"applatix.io/axops/commit"
	"applatix.io/axops/notification"
	"applatix.io/axops/service"
	"applatix.io/template"
)

type Deployment struct {
	Base

	TaskID                string   `json:"task_id,omitempty"`
	ApplicationGeneration string   `json:"app_generation,omitempty"`
	ApplicationID         string   `json:"app_id,omitempty"`
	ApplicationName       string   `json:"app_name,omitempty"`
	DeploymentID          string   `json:"deployment_id,omitempty"`
	JiraIssues            []string `json:"jira_issues"`

	PreviousDeploymentId string                            `json:"previous_deployment_id,omitempty"`
	Fixtures             map[string]map[string]interface{} `json:"fixtures,omitempty"`
	Instances            []*Pod                            `json:"instances"`
	Endpoints            []string                          `json:"endpoints"`
}

type Base struct {
	Id                string                              `json:"id,omitempty"`
	Name              string                              `json:"name,omitempty"`
	Description       string                              `json:"description,omitempty"`
	CostId            map[string]interface{}              `json:"costid,omitempty"`
	Template          *service.EmbeddedDeploymentTemplate `json:"template,omitempty"`
	TemplateID        string                              `json:"template_id,omitempty"`
	Arguments         template.Arguments                  `json:"arguments,omitempty"`
	Status            string                              `json:"status"`
	StatusDetail      map[string]interface{}              `json:"status_detail,omitempty"`
	Mem               float64                             `json:"mem,omitempty"`
	CPU               float64                             `json:"cpu,omitempty"`
	Cost              float64                             `json:"cost"`
	User              string                              `json:"user,omitempty"`
	Notifications     []notification.Notification         `json:"notifications,ommitempty"`
	CreateTime        int64                               `json:"create_time"`
	LaunchTime        int64                               `json:"launch_time"`
	EndTime           int64                               `json:"end_time"`
	WaitTime          int64                               `json:"wait_time"`
	RunTime           int64                               `json:"run_time"`
	Commit            *commit.ApiCommit                   `json:"commit,ommitempty"`
	Labels            map[string]string                   `json:"labels"`
	Annotations       map[string]string                   `json:"annotations"`
	TerminationPolicy *template.TerminationPolicy         `json:"termination_policy,omitempty"`
}

// String is the string representation of the deployment
func (d *Deployment) String() string {
	return fmt.Sprintf("Deployment %s (ID: %s) (DeploymentID: %s)", d.Name, d.Id, d.DeploymentID)
}

func (d *Deployment) Key() string {
	return d.DeploymentID
}

func (d *Deployment) HeartBeatKey() string {
	return d.Id
}

func GetLatestDeployments(params map[string]interface{}, external bool) ([]*Deployment, *axerror.AXError) {
	return getObjects(DeploymentLatestTable, params, external)
}

func GetHistoryDeployments(params map[string]interface{}, external bool) ([]*Deployment, *axerror.AXError) {
	return getObjects(DeploymentHistoryTable, params, external)
}

func GetLatestDeploymentByID(id string, external bool) (*Deployment, *axerror.AXError) {
	deployments, axErr := getObjects(
		DeploymentLatestTable,
		map[string]interface{}{
			axdb.AXDBUUIDColumnName: id,
		},
		external,
	)

	if axErr != nil {
		return nil, axErr
	}

	if len(deployments) == 0 {
		return nil, nil
	}

	d := deployments[0]
	return d, nil
}

func GetLatestDeploymentByName(appName, name string, external bool) (*Deployment, *axerror.AXError) {
	deployments, axErr := getObjects(
		DeploymentLatestTable,
		map[string]interface{}{
			ServiceName:       name,
			DeploymentAppName: appName,
		},
		external,
	)

	if axErr != nil {
		return nil, axErr
	}

	if len(deployments) == 0 {
		return nil, nil
	}

	d := deployments[0]
	return d, nil
}

func GetHistoryDeploymentByID(id string, external bool) (*Deployment, *axerror.AXError) {
	deployments, axErr := getObjects(
		DeploymentHistoryTable,
		map[string]interface{}{
			axdb.AXDBUUIDColumnName: id,
		},
		external,
	)

	if axErr != nil {
		return nil, axErr
	}

	if len(deployments) == 0 {
		return nil, nil
	}

	d := deployments[0]
	return d, nil
}

func GetDeploymentByID(id string, external bool) (*Deployment, *axerror.AXError) {
	d, axErr := GetLatestDeploymentByID(id, external)
	if axErr != nil {
		return nil, axErr
	}

	if d != nil {
		return d, nil
	}

	return GetHistoryDeploymentByID(id, external)
}

func GetLatestDeploymentsByApplication(app string, external bool) ([]*Deployment, *axerror.AXError) {
	deployments, axErr := getObjects(
		DeploymentLatestTable,
		map[string]interface{}{
			DeploymentAppName: app,
		},
		external,
	)

	if axErr != nil {
		return nil, axErr
	}

	return deployments, nil
}
